### PR TITLE
Convert hashSets in parallel before merge

### PR DESCRIPTION
**Pull request by jiebinn**
Before merge, if one of the lhs and rhs is singleLevelSet and the other is twoLevelSet, then the SingleLevelSet will call convertToTwoLevel(). The convert process is serial and not in parallel. It will cost lots of cycle before it cosumes all the singleLevelSet.

The idea of the patch is to convert all the singleLevelSets to twoLevelSets in parallel before merge.

I have tested the patch on Intel 2 x 112 vCPUs SPR server with clickbench and latest upstream ClickHouse. Q5 has got a big **264%** performance improvement and 24 queries have got at least 5% performance gain. The overall geomean of 43 queries has gained **7.4%** more than the base code.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This patch will provide a method to deal with all the hashsets in parallel before merge. 

### The detail to define the performance issue and how to resolve it.
First, we have found that there is a performance drop for the Q5 of clickbench as the core count increases.
image
Then we have collected a pipeline visualization if the max_threads of thread_pool is 80 or 112.
max_threads = 80
image
max_threads = 112
image
If the max_threads increase from 80 to 112, the merge stage does not decrease, but merge time is 3.2x more.
when merging two twoLevelHash, there is already an optimization to start thread_pool and merge in parallel. However, when merge one singleLevelHash and one TwoLevelHash, the singleLevelHash has to convert to twoLevelHash. The convert progress is serial.
If there is at least one singleLevelHash and one TwoLevelHash, all the singleLevelHash have to be converted to twoLevelHash before merged with the other twoLevelHash. We could add a new stage before merge called Prepare_Hash_before_Merge, where all the hashset would be processed before merging. All the singleLevelHash would be converted to twoLevelHash in parallel in this stage instead of in serial in merge stage.
with this patch, Q5 has got 2.64x performance improvement on a 2x112 vCPUs system (max_threads = 112).
image

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 38545463aade2f3019b81484d7324f92dcbd4ec2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50748/38545463aade2f3019b81484d7324f92dcbd4ec2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by nickitat** [comment]
pls take a look

```
Jun 11 02:52:18 FAILED: src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o 
Jun 11 02:52:18 /usr/bin/cmake -E __run_co_compile --launcher="prlimit;--as=10000000000;--data=5000000000;--cpu=1000;/usr/bin/sccache" --tidy="/usr/bin/clang-tidy-cache;/usr/bin/clang-tidy-16;--extra-arg-before=--driver-mode=g++" --source=/build/src/Interpreters/examples/two_level_hash_map.cpp -- /usr/bin/clang++-16 --target=x86_64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DANNOYLIB_MULTITHREADED_BUILD -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=10 -DAWS_SDK_VERSION_PATCH=36 -DBOOST_ASIO_HAS_STD_INVOKE_RESULT=1 -DBOOST_ASIO_STANDALONE=1 -DCARES_STATICLIB -DDUMMY_BACKTRACE -DENABLE_ANNOY -DENABLE_MULTITARGET_CODE=1 -DENABLE_QPL_COMPRESSION -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/build/build_docker/includes/configs -I/build/base/glibc-compatibility/memcpy -I/build/src -I/build/build_docker/src -I/build/build_docker/src/Core/include -I/build/base/base/.. -I/build/build_docker/base/base/.. -I/build/contrib/cctz/include -I/build/base/pcg-random/. -I/build/contrib/libfiu/libfiu -I/build/contrib/miniselect/include -I/build/contrib/zstd/lib -I/build/src/Common/mysqlxx/. -I/build/build_docker/rust/skim/include -isystem /build/contrib/llvm-project/libcxx/include -isystem /build/contrib/llvm-project/libcxxabi/include -isystem /build/contrib/libunwind/include -isystem /build/build_docker/contrib/orc/c++/include -isystem /build/contrib/llvm-project/llvm/include -isystem /build/build_docker/contrib/llvm-project/llvm/include -isystem /build/contrib/abseil-cpp -isystem /build/contrib/croaring/cpp -isystem /build/contrib/croaring/include -isystem /build/contrib/sparsehash-c11 -isystem /build/contrib/cityhash102/include -isystem /build/contrib/boost -isystem /build/base/poco/Net/include -isystem /build/base/poco/Foundation/include -isystem /build/base/poco/NetSSL_OpenSSL/include -isystem /build/base/poco/Crypto/include -isystem /build/contrib/boringssl/include -isystem /build/base/poco/Util/include -isystem /build/base/poco/JSON/include -isystem /build/base/poco/XML/include -isystem /build/contrib/replxx/include -isystem /build/contrib/fmtlib-cmake/../fmtlib/include -isystem /build/contrib/magic_enum/include -isystem /build/contrib/double-conversion -isystem /build/contrib/dragonbox/include -isystem /build/contrib/re2 -isystem /build/build_docker/contrib/re2-cmake -isystem /build/contrib/zlib-ng -isystem /build/build_docker/contrib/zlib-ng-cmake -isystem /build/contrib/pdqsort -isystem /build/contrib/xz/src/liblzma/api -isystem /build/contrib/aws/src/aws-cpp-sdk-core/include -isystem /build/build_docker/contrib/aws-cmake/include -isystem /build/contrib/aws/generated/src/aws-cpp-sdk-s3/include -isystem /build/contrib/aws-c-auth/include -isystem /build/contrib/aws-c-common/include -isystem /build/contrib/aws-c-io/include -isystem /build/contrib/aws-crt-cpp/include -isystem /build/contrib/aws-c-mqtt/include -isystem /build/contrib/aws-c-sdkutils/include -isystem /build/contrib/snappy -isystem /build/build_docker/contrib/snappy-cmake -isystem /build/contrib/libbcrypt -isystem /build/contrib/msgpack-c/include -isystem /build/build_docker/contrib/liburing/src/include-compat -isystem /build/build_docker/contrib/liburing/src/include -isystem /build/contrib/liburing/src/include -isystem /build/contrib/fast_float/include -isystem /build/contrib/NuRaft/include -isystem /build/base/poco/MongoDB/include -isystem /build/build_docker/contrib/mariadb-connector-c-cmake/include-public -isystem /build/contrib/mariadb-connector-c/include -isystem /build/contrib/mariadb-connector-c/libmariadb -isystem /build/build_docker/src/Server/grpc_protos -isystem /build/contrib/grpc/include -isystem /build/contrib/c-ares/src/lib -isystem /build/contrib/c-ares/include -isystem /build/contrib/c-ares-cmake/linux -isystem /build/contrib/google-protobuf/src -isystem /build/contrib/s2geometry/src -isystem /build/contrib/AMQP-CPP/include -isystem /build/contrib/AMQP-CPP -isystem /build/contrib/libuv/include -isystem /build/contrib/sqlite-amalgamation -isystem /build/contrib/rocksdb/include -isystem /build/contrib/libpqxx/include -isystem /build/contrib/libpq -isystem /build/contrib/libpq/include -isystem /build/contrib/qpl-cmake -isystem /build/contrib/qpl/include -isystem /build/contrib/idxd-config/accfg -isystem /build/contrib/libstemmer_c/include -isystem /build/contrib/wordnet-blast -isystem /build/contrib/lemmagen-c/include -isystem /build/contrib/ulid-c/include -isystem /build/contrib/consistent-hashing -isystem /build/contrib/annoy/src --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -UNDEBUG -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -g -O0 -g -gdwarf-4  -D_LIBCPP_DEBUG=0 -std=c++2b   -D OS_LINUX -I/build/base -I/build/contrib/magic_enum/include -include /build/src/Core/iostream_debug_helpers.h -Werror -nostdinc++ -MD -MT src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -MF src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o.d -o src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -c /build/src/Interpreters/examples/two_level_hash_map.cpp
Jun 11 02:52:18 /build/src/AggregateFunctions/UniquesHashSet.h:369:91: error: unknown type name 'ThreadPool' [clang-diagnostic-error]
Jun 11 02:52:18     static void parallelizeMergePrepare(const std::vector<UniquesHashSet *> & /*places*/, ThreadPool * /*thread_pool = nullptr*/) {}
```

File: src/Interpreters/Aggregator.cpp [hunk 1]
```diff
@@ -2601,6 +2601,21 @@ void NO_INLINE Aggregator::mergeWithoutKeyDataImpl(
 
     AggregatedDataVariantsPtr & res = non_empty_data[0];
 
+    for (size_t i = 0; i < params.aggregates_size; ++i)
+    {
+        if (aggregate_functions[i]->isAbleToParallelizeMerge() &&
```
Comment (nickitat) [inline src/Interpreters/Aggregator.cpp:2606]: `isAbleToParallelizeMerge` looks irrelevant here, `isParallelizeMergePrepareNeeded` should be enough

File: src/AggregateFunctions/UniqExactSet.h [hunk 2]
```diff
@@ -27,6 +27,56 @@ class UniqExactSet
             asTwoLevel().insert(std::forward<Arg>(arg));
     }
 
+    /// In merge, if one of the lhs and rhs is twolevelset and the other is singlelevelset, then the singlelevelset will need to convertToTwoLevel().
+    /// It's not in parallel and will cost extra large time if the thread_num is large.
+    /// This method will convert all the SingleLevelSet to TwoLevelSet in parallel if the hashsets are not all singlelevel or not all twolevel.
+    static void parallelizeMergePrepare(const std::vector<UniqExactSet *> & data_vec, ThreadPool * thread_pool = nullptr)
```
Comment (nickitat) [inline src/AggregateFunctions/UniqExactSet.h:33]: could we make sure that `ThreadPool` is always not null here and pass it as a reference?

File: src/AggregateFunctions/UniqExactSet.h [hunk 3]
```diff
@@ -27,6 +27,56 @@ class UniqExactSet
...
+                    {
+                        const auto i = data_vec_atomic_index->fetch_add(1);
+                        if (i >= data_vec.size())
+                            return;
+                        if (data_vec[i]->isSingleLevel())
+                            data_vec[i]->convertToTwoLevel();
+                    }
+                };
+                for (size_t i = 0; i < std::min<size_t>(thread_pool->getMaxThreads(), single_level_set_num); ++i)
+                    thread_pool->scheduleOrThrowOnError(thread_func);
```
Comment (nickitat) [inline src/AggregateFunctions/UniqExactSet.h:73]: if exception will be thrown here, we will not reach `wait()`. all this code should be wrapped into try-catch that also does `wait` in case of exception (refer to #50590).

**Comment by nickitat** [comment]
would be good to find perf-tests that shows speed-up in the addressed case.

File: src/AggregateFunctions/AggregateFunctionUniq.h [hunk 4]
```diff
@@ -408,6 +409,17 @@ class AggregateFunctionUniq final : public IAggregateFunctionDataHelper<Data, Ag
         detail::Adder<T, Data>::add(this->data(place), columns, num_args, row_begin, row_end, flags, null_map);
     }
 
+    bool isParallelizeMergePrepareNeeded() const override { return true;}
+
+    void parallelizeMergePrepare(AggregateDataPtrs & places, ThreadPool & thread_pool) const override
+    {
+        std::vector<DataSet *> data_vec;
```
Comment (devcrafter) [inline src/AggregateFunctions/AggregateFunctionUniq.h:416]:
```
data_vec.resize(places.size());
```

**Comment by jiebinn** [comment]
@nickitat @devcrafter Thanks for your kind review. I will think about that and let you know once I finish.

**Comment by jiebinn** [comment]
> pls take a look
> 
> ```
> Jun 11 02:52:18 FAILED: src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o 
> Jun 11 02:52:18 /usr/bin/cmake -E __run_co_compile --launcher="prlimit;--as=10000000000;--data=5000000000;--cpu=1000;/usr/bin/sccache" --tidy="/usr/bin/clang-tidy-cache;/usr/bin/clang-tidy-16;--extra-arg-before=--driver-mode=g++" --source=/build/src/Interpreters/examples/two_level_hash_map.cpp -- /usr/bin/clang++-16 --target=x86_64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DANNOYLIB_MULTITHREADED_BUILD -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=10 -DAWS_SDK_VERSION_PATCH=36 -DBOOST_ASIO_HAS_STD_INVOKE_RESULT=1 -DBOOST_ASIO_STANDALONE=1 -DCARES_STATICLIB -DDUMMY_BACKTRACE -DENABLE_ANNOY -DENABLE_MULTITARGET_CODE=1 -DENABLE_QPL_COMPRESSION -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/build/build_docker/includes/configs -I/build/base/glibc-compatibility/memcpy -I/build/src -I/build/build_docker/src -I/build/build_docker/src/Core/include -I/build/base/base/.. -I/build/build_docker/base/base/.. -I/build/contrib/cctz/include -I/build/base/pcg-random/. -I/build/contrib/libfiu/libfiu -I/build/contrib/miniselect/include -I/build/contrib/zstd/lib -I/build/src/Common/mysqlxx/. -I/build/build_docker/rust/skim/include -isystem /build/contrib/llvm-project/libcxx/include -isystem /build/contrib/llvm-project/libcxxabi/include -isystem /build/contrib/libunwind/include -isystem /build/build_docker/contrib/orc/c++/include -isystem /build/contrib/llvm-project/llvm/include -isystem /build/build_docker/contrib/llvm-project/llvm/include -isystem /build/contrib/abseil-cpp -isystem /build/contrib/croaring/cpp -isystem /build/contrib/croaring/include -isystem /build/contrib/sparsehash-c11 -isystem /build/contrib/cityhash102/include -isystem /build/contrib/boost -isystem /build/base/poco/Net/include -isystem /build/base/poco/Foundation/include -isystem /build/base/poco/NetSSL_OpenSSL/include -isystem /build/base/poco/Crypto/include -isystem /build/contrib/boringssl/include -isystem /build/base/poco/Util/include -isystem /build/base/poco/JSON/include -isystem /build/base/poco/XML/include -isystem /build/contrib/replxx/include -isystem /build/contrib/fmtlib-cmake/../fmtlib/include -isystem /build/contrib/magic_enum/include -isystem /build/contrib/double-conversion -isystem /build/contrib/dragonbox/include -isystem /build/contrib/re2 -isystem /build/build_docker/contrib/re2-cmake -isystem /build/contrib/zlib-ng -isystem /build/build_docker/contrib/zlib-ng-cmake -isystem /build/contrib/pdqsort -isystem /build/contrib/xz/src/liblzma/api -isystem /build/contrib/aws/src/aws-cpp-sdk-core/include -isystem /build/build_docker/contrib/aws-cmake/include -isystem /build/contrib/aws/generated/src/aws-cpp-sdk-s3/include -isystem /build/contrib/aws-c-auth/include -isystem /build/contrib/aws-c-common/include -isystem /build/contrib/aws-c-io/include -isystem /build/contrib/aws-crt-cpp/include -isystem /build/contrib/aws-c-mqtt/include -isystem /build/contrib/aws-c-sdkutils/include -isystem /build/contrib/snappy -isystem /build/build_docker/contrib/snappy-cmake -isystem /build/contrib/libbcrypt -isystem /build/contrib/msgpack-c/include -isystem /build/build_docker/contrib/liburing/src/include-compat -isystem /build/build_docker/contrib/liburing/src/include -isystem /build/contrib/liburing/src/include -isystem /build/contrib/fast_float/include -isystem /build/contrib/NuRaft/include -isystem /build/base/poco/MongoDB/include -isystem /build/build_docker/contrib/mariadb-connector-c-cmake/include-public -isystem /build/contrib/mariadb-connector-c/include -isystem /build/contrib/mariadb-connector-c/libmariadb -isystem /build/build_docker/src/Server/grpc_protos -isystem /build/contrib/grpc/include -isystem /build/contrib/c-ares/src/lib -isystem /build/contrib/c-ares/include -isystem /build/contrib/c-ares-cmake/linux -isystem /build/contrib/google-protobuf/src -isystem /build/contrib/s2geometry/src -isystem /build/contrib/AMQP-CPP/include -isystem /build/contrib/AMQP-CPP -isystem /build/contrib/libuv/include -isystem /build/contrib/sqlite-amalgamation -isystem /build/contrib/rocksdb/include -isystem /build/contrib/libpqxx/include -isystem /build/contrib/libpq -isystem /build/contrib/libpq/include -isystem /build/contrib/qpl-cmake -isystem /build/contrib/qpl/include -isystem /build/contrib/idxd-config/accfg -isystem /build/contrib/libstemmer_c/include -isystem /build/contrib/wordnet-blast -isystem /build/contrib/lemmagen-c/include -isystem /build/contrib/ulid-c/include -isystem /build/contrib/consistent-hashing -isystem /build/contrib/annoy/src --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -UNDEBUG -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -g -O0 -g -gdwarf-4  -D_LIBCPP_DEBUG=0 -std=c++2b   -D OS_LINUX -I/build/base -I/build/contrib/magic_enum/include -include /build/src/Core/iostream_debug_helpers.h -Werror -nostdinc++ -MD -MT src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -MF src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o.d -o src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -c /build/src/Interpreters/examples/two_level_hash_map.cpp
> Jun 11 02:52:18 /build/src/AggregateFunctions/UniquesHashSet.h:369:91: error: unknown type name 'ThreadPool' [clang-diagnostic-error]
> Jun 11 02:52:18     static void parallelizeMergePrepare(const std::vector<UniquesHashSet *> & /*places*/, ThreadPool * /*thread_pool = nullptr*/) {}
> ```

The latest code has involved the head file '<Common/ThreadPool_fwd.h>' in data set, such as 'ThetaSketchData', 'HyperLogLogWithSmallSetOptimization' and 'UniquesHashSet'.

File: src/AggregateFunctions/IAggregateFunction.h [hunk 5]
```diff
@@ -147,6 +148,10 @@ class IAggregateFunction : public std::enable_shared_from_this<IAggregateFunctio
     /// Default values must be a the 0-th positions in columns.
     virtual void addManyDefaults(AggregateDataPtr __restrict place, const IColumn ** columns, size_t length, Arena * arena) const = 0;
 
+    virtual bool isParallelizeMergePrepareNeeded() const { return false; }
+
+    virtual void parallelizeMergePrepare(AggregateDataPtrs & /*places*/, ThreadPool & /*thread_pool*/) const {}
```
[the remark this answers is not in the document]
Comment (jiebinn) [reply src/AggregateFunctions/IAggregateFunction.h:153]: Done. Add the exception in virtual function.

File: src/AggregateFunctions/AggregateFunctionUniq.h [hunk 6]
```diff
@@ -408,6 +409,17 @@ class AggregateFunctionUniq final : public IAggregateFunctionDataHelper<Data, Ag
         detail::Adder<T, Data>::add(this->data(place), columns, num_args, row_begin, row_end, flags, null_map);
     }
 
+    bool isParallelizeMergePrepareNeeded() const override { return true;}
+
+    void parallelizeMergePrepare(AggregateDataPtrs & places, ThreadPool & thread_pool) const override
```
[the remark this answers is not in the document]
Comment (jiebinn) [reply src/AggregateFunctions/AggregateFunctionUniq.h:414]: Yes. The `is_able_to_parallelize_merge` is unique for each data set. Then I think it could be reused in UniqExactSet. Update it in latest code.

File: src/Interpreters/Aggregator.cpp (hunk 1 above)
Comment (jiebinn) [reply src/Interpreters/Aggregator.cpp:2606]: Keep the `isAbleToParallelizeMerge` and delete 'isParallelizeMergePrepareNeeded'.

File: src/AggregateFunctions/UniqExactSet.h (hunk 2 above)
Comment (jiebinn) [reply src/AggregateFunctions/UniqExactSet.h:33]: From the `void parallelizeMergePrepare(AggregateDataPtrs & places, ThreadPool & thread_pool) const override` in `AggregateFunctionUniq.h`, the `ThreadPool` is alway not null. The latest code has used the reference instead of a pointer.

File: src/AggregateFunctions/UniqExactSet.h (hunk 3 above)
Comment (jiebinn) [reply src/AggregateFunctions/UniqExactSet.h]: Add try catch.

**Comment by jiebinn** [comment]
> would be good to find perf-tests that shows speed-up in the addressed case.

I have used the clickbench for the performance test. You mean we might find the "distinct/uniq" operation in tests and check the performance?

File: src/AggregateFunctions/AggregateFunctionUniq.h (hunk 4 above)
Comment (jiebinn) [reply src/AggregateFunctions/AggregateFunctionUniq.h:414]: I have tried to add the code line `data_vec.resize(places.size());` after the initialization of `std::vector<DataSet *> data_vec;` But the ClickHouse server crashed. I will check that.

File: src/AggregateFunctions/AggregateFunctionUniq.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiebinn) [reply src/AggregateFunctions/AggregateFunctionUniq.h:426]: The `is_able_to_parallelize_merge` could be true or false in different data sets, such as `HyperLogLogWithSmallSetOptimization`, where the merge is not parallel. The optimization patch is only for the data set 'UniqExactSet'. Then only the flag `is_able_to_parallelize_merge` is not enough for `parallelizeMergePrepare`. And we have to add another `is_parallelize_merge_prepare_needed`.

File: src/Interpreters/Aggregator.cpp (hunk 1 above)
Comment (jiebinn) [reply src/Interpreters/Aggregator.cpp]: The flag `isAbleToParallelizeMerge` is for the parallel merge of UniqExact. If the merge stage is not parallel (`isAbleToParallelizeMerge = false`), we could still do hashset converting in parallel. The merge stage and hashset converting stage are independent and irrelevant. Then I think `isAbleToParallelizeMerge` could be removed.

**Comment by jiebinn** [comment]
Hi @nickitat @devcrafter, thanks for your previous code review. I have updated the patch code and comments according to the review of PR.

File: src/AggregateFunctions/AggregateFunctionUniq.h (hunk 4 above)
Comment (jiebinn) [reply src/AggregateFunctions/AggregateFunctionUniq.h:461]: Add the resize() for data_vec in parallelizeMergePrepare.

**Comment by jiebinn** [comment]
@nickitat The CI error has succeeded in the latest attempt. But the status doesn't show that.
image

**Comment by jiebinn** [comment]
Update the `clickbench Q5 performance test of core scaling` and `pipeline` figures, which would show how to define the issue and make the `merge` more scalable.

**Comment by jiebinn** [comment]
@nickitat @devcrafter @kitaisreal
I have added the performance test for the data set hits_v1, which have 2.1x and 1.8x performance improvement on my machine.

<query>SELECT COUNT(DISTINCT Title) FROM hits_v1 SETTINGS max_threads = 24</query>
<query>SELECT COUNT(DISTINCT Referer) FROM hits_v1 SETTINGS max_threads = 22</query>

**Comment by jiebinn** [comment]
Rename the data set name of performance test from hits_v1 to test.hits to fit the CI rule.

File: src/AggregateFunctions/ThetaSketchData.h [hunk 7]
```diff
@@ -66,6 +66,8 @@ class ThetaSketchData : private boost::noncopyable
             return 0;
     }
 
+    static void parallelizeMergePrepare(const std::vector<ThetaSketchData *> & /*places*/, ThreadPool & /*thread_pool*/) {}
```
Comment (nickitat) [inline src/AggregateFunctions/ThetaSketchData.h:69]: if I'm not missing smth all `parallelizeMergePrepare` could be one of:
* not defined/declared at all
* throw exception
* do meaningful work (empty implementation is not meaningful)

I have some prejudice towards empty methods, what do you think?

File: src/AggregateFunctions/UniqExactSet.h [hunk 8]
```diff
@@ -28,6 +28,61 @@ class UniqExactSet
...
+        for (auto ele : data_vec)
+        {
+            if (ele->isSingleLevel())
+                single_level_set_num ++;
+        }
+
+        if (single_level_set_num > 0 && single_level_set_num < data_vec.size())
+            is_convert_needed = true;
+
+        if (is_convert_needed)
```
Comment (nickitat) [inline src/AggregateFunctions/UniqExactSet.h:48]:
```suggestion
```

File: src/AggregateFunctions/ThetaSketchData.h (hunk 7 above)
Comment (jiebinn) [reply src/AggregateFunctions/ThetaSketchData.h]: Agree. 

- For these empty methods in other datasets, we could add the throw exception. 

- Or we may just delete these empty methods in other datasets. Call the `DataSet::parallelizeMergePrepare()` only when `is_parallelize_merge_prepare_needed` is true.

I would like to suggest the 2nd way.

**Review by nickitat — APPROVED** [review]

**Comment by jiebinn** [comment]
The failing check - stress_test (msan) is successful in latest attempt. CI does not update the latest status on the webpage.
image
